packages/domain/src/amount.ts

```diff
@@ -0,0 +1,186 @@
+import { ONE, TEN, TEN_POW_18, TYPICAL_PRECISION, ZERO } from '@dma-common/constants'
+import BigNumber from 'bignumber.js'
+
+/**
+ * Possible precision levels are:
+ * - none                                             EG 1 USDC = 1
+ * - (no precision set, just the raw amount)
+ *
+ * - tokenMax                                                EG 1 USDC = 1e6
+ * - (max precision for the token)
+ *
+ * - normalized                                         EG 1 USDC = 1e18
+ * - (precision set 18 decimals for to standardise amounts for maths reasons)
+ */
+type PrecisionMode = 'none' | 'tokenMax' | 'normalized'
+type PrecisionMap = Record<PrecisionMode, BigNumber>
+
+type AmountArgs = {
+  amount: BigNumber
+  precision?: {
+    /** none: 1USDC = 1, max: 1USDC = 1e6, normalized: 1USDC = 1e18 */
+    mode?: PrecisionMode
+    tokenMaxDecimals?: number
+  }
+}
+
+export class Amount {
+  private currentPrecisionScale: BigNumber
+  private readonly tokenMaxDecimals: number
+  private amount: BigNumber
+  private readonly precisionMap: PrecisionMap
+  private currentPrecisionMode: 'none' | 'tokenMax' | 'normalized' = 'none'
+
+  constructor(args: AmountArgs) {
+    const initialAmount = args.amount || ZERO
+    const initialPrecisionMode = args?.precision?.mode || 'none'
+    this.tokenMaxDecimals = args?.precision?.tokenMaxDecimals || TYPICAL_PRECISION
+    this.precisionMap = {
+      none: ONE,
+      tokenMax: new BigNumber(TEN.pow(this.tokenMaxDecimals)),
```

Comment:
I'm not a fan of the name `tokenMax`. A big number is usually used to represent a large quantity, not a balance. It could be a percentage. It's like creating a wrapper for a float number and passing a currency symbol as a parameter.